File: examples/kotlin/VulnerableController.kt
import java.net.URL
import java.security.MessageDigest
import java.util.Random

class VulnerableController {
    fun run(user: String, url: String, cmd: String, response: javax.servlet.http.HttpServletResponse) {
        val query = "SELECT * FROM users WHERE name = '" + user + "'"
        println(query)

        response.writer.write("<div>" + user + "</div>")
        Runtime.getRuntime().exec("sh -c " + cmd)

        URL(url).openConnection()

        val digest = MessageDigest.getInstance("MD5")
        val token = Random().nextInt()
        val apiKey = "super-secret-key"

        println(digest)
        println(token)
        println(apiKey)
    }
}

File: examples/csharp/VulnerableController.cs
using System;
using System.Data.SqlClient;
using System.Diagnostics;
using System.Net.Http;
using System.Security.Cryptography;
using Microsoft.AspNetCore.Mvc;

namespace Demo.Controllers
{
    public class VulnerableController : Controller
    {
        public void Run(string user, string url, string cmd)
        {
            string query = "SELECT * FROM Users WHERE Name = '" + user + "'";
            var command = new SqlCommand(query);

            Response.Write("<div>" + user + "</div>");
            Process.Start("cmd.exe", cmd);

            var client = new HttpClient();
            client.GetStringAsync(url);

            var md5 = MD5.Create();
            var rng = new Random();

            string password = "demo-insecure-password";
            Console.WriteLine(password);
        }
    }
}
